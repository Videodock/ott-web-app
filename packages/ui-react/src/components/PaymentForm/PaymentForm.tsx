import { useEffect } from 'react';
import { useTranslation } from 'react-i18next';
import { useLocation, useNavigate } from 'react-router-dom';
import Payment from 'payment';
import { object, string } from 'yup';
import { getModule } from '@jwp/ott-common/src/modules/container';
import CheckoutController from '@jwp/ott-common/src/stores/CheckoutController';
import useForm from '@jwp/ott-hooks-react/src/useForm';
import useCheckAccess from '@jwp/ott-hooks-react/src/useCheckAccess';
import { createURL } from '@jwp/ott-common/src/utils/urlFormatting';

import Button from '../Button/Button';
import CreditCardCVCField from '../CreditCardCVCField/CreditCardCVCField';
import CreditCardExpiryField from '../CreditCardExpiryField/CreditCardExpiryField';
import CreditCardNumberField from '../CreditCardNumberField/CreditCardNumberField';
import TextField from '../TextField/TextField';
import { modalURLFromLocation } from '../../utils/location';
import { useAriaAnnouncer } from '../../containers/AnnouncementProvider/AnnoucementProvider';

import styles from './PaymentForm.module.scss';

type Props = {
  couponCode?: string;
  setUpdatingOrder: (value: boolean) => void;
};

const PaymentForm: React.FC<Props> = ({ couponCode, setUpdatingOrder }) => {
  const checkoutController = getModule(CheckoutController);
  const announcer = useAriaAnnouncer();

  const { t } = useTranslation('account');
  const location = useLocation();
  const navigate = useNavigate();
  const { intervalCheckAccess } = useCheckAccess();

  const paymentData = useForm(
    { cardholderName: '', cardNumber: '', cardExpiry: '', cardCVC: '', cardExpMonth: '', cardExpYear: '' },
    async () => {
      setUpdatingOrder(true);

      const referrer = window.location.href;
      const returnUrl = createURL(window.location.href, { u: 'waiting-for-payment' });

      await checkoutController.directPostCardPayment({ couponCode, ...paymentData.values }, referrer, returnUrl);

      intervalCheckAccess({
        interval: 15000,
        callback: (hasAccess) => {
          if (hasAccess) {
            announcer(t('checkout.payment_success'), 'success');
            navigate(modalURLFromLocation(location, 'welcome'));
          }
        },
      });
    },
    object().shape({
      cardNumber: string().test('card number validation', t('checkout.invalid_card_number'), (value) => {
        return Payment.fns.validateCardNumber(value as string);
      }),
      cardExpiry: string().test('card expiry validation', t('checkout.invalid_card_expiry'), (value) => {
        return Payment.fns.validateCardExpiry(value as string);
      }),
      cardCVC: string().test('cvc validation', t('checkout.invalid_cvc_number'), (value) => {
        const issuer = Payment.fns.cardType(paymentData?.values?.cardNumber);
        return Payment.fns.validateCardCVC(value as string, issuer);
      }),
    }),
    true,
  );

  useEffect(() => {
    if (paymentData.values.cardExpiry) {
      const expiry = Payment.fns.cardExpiryVal(paymentData.values.cardExpiry);
      if (expiry.month) {
        paymentData.setValue('cardExpMonth', expiry.month.toString());
      }
      if (expiry.year) {
        paymentData.setValue('cardExpYear', expiry.year.toString());
      }
    }
    //eslint-disable-next-line
  }, [paymentData.values.cardExpiry]);

  return (
    <div className={styles.paymentForm}>
      <div>
        <TextField
          label={t('checkout.card_holder_name')}
          name="cardholderName"
          value={paymentData?.values?.cardholderName}
          onChange={paymentData?.handleChange}
          onBlur={paymentData?.handleBlur}
          placeholder={t('checkout.credit_card_name')}
          required
        />
      </div>
      <div>
        <CreditCardNumberField
          value={paymentData?.values?.cardNumber?.toString()}
          error={paymentData?.errors?.cardNumber}
          onBlur={paymentData?.handleBlur}
          onChange={paymentData?.handleChange}
        />
      </div>
      <div className={styles.columns}>
        <div>
          <CreditCardExpiryField
            value={paymentData?.values?.cardExpiry}
            onBlur={paymentData?.handleBlur}
            onChange={paymentData?.handleChange}
            error={paymentData?.errors?.cardExpiry}
          />
        </div>
        <div>
          <CreditCardCVCField
            value={paymentData?.values?.cardCVC}
            onBlur={paymentData?.handleBlur}
            onChange={paymentData?.handleChange}
            error={paymentData?.errors?.cardCVC}
          />
        </div>
      </div>
      <div>
        <Button label={t('checkout.continue')} variant="contained" color="primary" onClick={paymentData.handleSubmit as () => void} size="large" fullWidth />
      </div>
    </div>
  );
};

export default PaymentForm;
