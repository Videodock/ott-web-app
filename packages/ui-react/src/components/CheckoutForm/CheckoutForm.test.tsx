import React from 'react';
import { render } from '@testing-library/react';
import type { Offer, Order } from '@jwp/ott-common/types/checkout';
import offer from '@jwp/ott-testing/fixtures/monthlyOffer.json';
import order from '@jwp/ott-testing/fixtures/order.json';

import PaymentForm from '../PaymentForm/PaymentForm';

import CheckoutForm from './CheckoutForm';

describe('<CheckoutForm>', () => {
  test('renders and matches snapshot', () => {
    const { container } = render(
      <CheckoutForm
        onCloseCouponFormClick={vi.fn()}
        onCouponFormSubmit={vi.fn()}
        onRedeemCouponButtonClick={vi.fn()}
        onCouponInputChange={vi.fn()}
        onBackButtonClick={vi.fn()}
        onPaymentMethodChange={vi.fn()}
        couponFormOpen={false}
        couponFormApplied={false}
        couponInputValue=""
        couponFormError={undefined}
        couponFormSubmitting={false}
        order={order as Order}
        offer={offer as Offer}
        offerType={'svod'}
        submitting={false}
      >
        <PaymentForm onPaymentFormSubmit={vi.fn()} />
      </CheckoutForm>,
    );

    expect(container).toMatchSnapshot();
  });
});
