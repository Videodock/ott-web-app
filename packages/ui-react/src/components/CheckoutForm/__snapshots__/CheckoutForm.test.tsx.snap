// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<CheckoutForm> > renders and matches snapshot 1`] = `
<div>
  <div>
    <h1
      class="_title_abc9bc"
    >
      checkout.payment_method
    </h1>
    <div
      class="_order_abc9bc"
    >
      <div
        class="_orderInfo_abc9bc"
      >
        <p
          class="_orderTitle_abc9bc _orderTitleMargin_abc9bc"
        >
          checkout.monthly
        </p>
      </div>
      <div
        class="_orderPrice_abc9bc"
      >
        <span>
          € 6,99
        </span>
        <small>
          /
          periods.month
        </small>
      </div>
    </div>
    <div
      class="_couponForm_abc9bc"
    >
      <button
        class="_button_f8f296 _default_f8f296 _outlined_f8f296"
        type="button"
      >
        <span>
          checkout.redeem_coupon
        </span>
      </button>
    </div>
    <div>
      <table
        class="_orderTotals_abc9bc"
      >
        <tbody />
        <tfoot>
          <tr>
            <td>
              checkout.total_price
            </td>
            <td>
              € 6,99
            </td>
          </tr>
          <tr>
            <td>
              checkout.applicable_tax
            </td>
            <td>
              € 1,21
            </td>
          </tr>
        </tfoot>
      </table>
    </div>
    <hr
      class="_divider_abc9bc"
    />
    <div
      class="_paymentMethods_abc9bc"
    />
    <div>
      <div
        class="_paymentForm_65d7c4"
      >
        <form
          data-testid="payment_form"
          novalidate=""
        >
          <div>
            <div
              class="_formField_86f2f5 _textField_c799b0"
            >
              <label
                class="_label_86f2f5"
                for="text-field_1235_cardholdername"
              >
                checkout.card_holder_name
                <span
                  aria-hidden="true"
                >
                  *
                </span>
              </label>
              <div
                class="_container_859b20"
              >
                <input
                  aria-describedby="helper_text_1235_cardholdername"
                  aria-invalid="false"
                  aria-required="true"
                  class="_input_859b20"
                  id="text-field_1235_cardholdername"
                  name="cardholderName"
                  placeholder="checkout.credit_card_name"
                  required=""
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <div>
            <div
              class="_formField_86f2f5 _textField_c799b0 directPostCreditCardNumber"
            >
              <label
                class="_label_86f2f5"
                for="text-field_1235_cardnumber"
              >
                Card number
                <span
                  aria-hidden="true"
                >
                  *
                </span>
              </label>
              <div
                class="_container_859b20"
              >
                <input
                  aria-describedby="helper_text_1235_cardnumber"
                  aria-invalid="false"
                  aria-required="true"
                  class="_input_859b20"
                  id="text-field_1235_cardnumber"
                  name="cardNumber"
                  pattern="\\\\d*"
                  placeholder="1234 5678 9012 3456"
                  required=""
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="_columns_65d7c4"
          >
            <div>
              <div
                class="_formField_86f2f5 _textField_c799b0 directPostExpiryDate"
              >
                <label
                  class="_label_86f2f5"
                  for="text-field_1235_cardexpiry"
                >
                  Expiry date
                  <span
                    aria-hidden="true"
                  >
                    *
                  </span>
                </label>
                <div
                  class="_container_859b20"
                >
                  <input
                    aria-describedby="helper_text_1235_cardexpiry"
                    aria-invalid="false"
                    aria-required="true"
                    class="_input_859b20"
                    id="text-field_1235_cardexpiry"
                    name="cardExpiry"
                    pattern="\\\\d*"
                    placeholder="MM/YY"
                    required=""
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div>
              <div
                class="_formField_86f2f5 _textField_c799b0 directPostSecurityCode"
              >
                <label
                  class="_label_86f2f5"
                  for="text-field_1235_cardcvc"
                >
                  payment.security_code
                  <span
                    aria-hidden="true"
                  >
                    *
                  </span>
                </label>
                <div
                  class="_container_859b20"
                >
                  <input
                    aria-describedby="helper_text_1235_cardcvc"
                    aria-invalid="false"
                    aria-label="payment.security_code"
                    aria-required="true"
                    class="_input_859b20"
                    id="text-field_1235_cardcvc"
                    name="cardCVC"
                    pattern="\\\\d*"
                    placeholder="###"
                    required=""
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div>
            <button
              class="_button_f8f296 _primary_f8f296 _fullWidth_f8f296 _large_f8f296"
              type="submit"
            >
              <span>
                checkout.continue
              </span>
            </button>
          </div>
        </form>
      </div>
    </div>
    <div
      aria-label="common:back"
      class="_iconButton_0fef65 _dialogBackButton_248199"
      role="button"
      tabindex="0"
    >
      <svg
        aria-hidden="true"
        class="_icon_585b29"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20,11V13H8L13.5,18.5L12.08,19.92L4.16,12L12.08,4.08L13.5,5.5L8,11H20Z"
        />
      </svg>
    </div>
  </div>
</div>
`;
